### PR TITLE
appveyor: enable building package

### PR DESCRIPTION
**Pull request by kenhys**
TODO: fix stalling  install-mroonga.ps1 issue

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -31,39 +43,77 @@ install:
...
+      - pacman -Sy --noconfirm bison
+      - ps: .\get-mroonga-batfiles.ps1
+      - ps: .\prepare-building-mroonga.ps1
+      - cmake --version
+      - bison --version
+      - move ..\..\mariadb-%MARIADB_VERSION% work/source
+      - cd .\work
+      - call .\build-vc2015-zip-64.bat
+      - cd ..\
+      - ps: .\package-unzip.ps1
```
Comment (cosmo0920) [inline appveyor.yml]: Could you use `7z x` for other archive extractor?
Currently, this posershell script loads configuration which specifies other MariaDB version.
So, it causes CI failure.

File: appveyor.yml [hunk 2]
```diff
@@ -31,39 +43,75 @@ install:
...
+      - ps: .\prepare-building-mroonga.ps1
+      - cmake --version
+      - bison --version
+      - move ..\..\mariadb-%MARIADB_VERSION% work/source
+      - cd .\work
+      - call .\build-vc2015-zip-64.bat
+      - 7z x mariadb-%MARIADB_VERSION%-winx64.zip
+      - cd ..\
+      - netsh advfirewall firewall add rule name="mariadb" dir=in action=allow protocol=TCP localport=3306 enable=yes profile=any
+      - ps: .\install-mroonga.ps1
```
Comment (cosmo0920) [inline appveyor.yml]: Also, install-mroonga.ps1 and create-mrnzip.ps1 use configuration which specifies other MariaDB version.
We should also write equivalent script.
Or, replace variables on [versions.ps1](https://github.com/cosmo0920/PowerShell-for-Mroonga-building/blob/master/powershell/versions.ps1#L3-L4).

File: appveyor.yml (hunk 1 above)
Comment (kenhys) [reply appveyor.yml]: Fixed

File: appveyor.yml (hunk 2 above)
Comment (kenhys) [reply appveyor.yml]: Fixed.

**Comment by kenhys** [comment]
@komainu8 Could you review this PR?

Now, package is available as a artifacat from https://ci.appveyor.com/project/groonga/mroonga/builds/22516277/job/4eo48sqmvj4oj6ti/artifacts

File: appveyor.yml [hunk 3]
```diff
@@ -27,43 +39,85 @@ install:
...
+        - configuration: Package
+    build_script:
+      - set
+      - cd %APPVEYOR_BUILD_FOLDER%\..
+      - git clone --depth 1
+        https://github.com/cosmo0920/PowerShell-for-Mroonga-building.git mrnbuild
+      - cd mrnbuild/powershell
+      - ps: $env:MARIADBVER = "mariadbVer = " + '"' + $env:MARIADB_VERSION + '"'
+      - ps: $env:MROONGAVER = "mroongaVer = " + '"' + $env:MROONGA_VERSION + '"'
+      - ps: (Get-Content .\versions.ps1) | ForEach-Object { $_ -replace 'mariadbVer = ".+"',$env:MARIADBVER } | ForEach-Object { $_ -replace 'mroongaVer = ".+"',$env:MROONGAVER } | Set-Content .\versions.ps1
```
Comment (cosmo0920) [inline appveyor.yml]: :+1:

File: appveyor.yml [hunk 4]
```diff
@@ -27,43 +39,85 @@ install:
...
+      - ps: .\prepare-building-mroonga.ps1
+      - cmake --version
+      - bison --version
+      - move ..\..\mariadb-%MARIADB_VERSION% work/source
+      - cd .\work
+      - call .\build-vc2015-zip-64.bat
+      - 7z x mariadb-%MARIADB_VERSION%-winx64.zip
+      - cd ..\
+      - netsh advfirewall firewall add rule name="mariadb" dir=in action=allow protocol=TCP localport=3306 enable=yes profile=any
+      - ps: (Get-Content .\install-mroonga.ps1) | ForEach-Object { $_ -replace '"win32", ','' } | Set-Content .\install-mroonga.ps1
```
Comment (cosmo0920) [inline appveyor.yml]: Whoa! It is intersting. :open_mouth:

File: appveyor.yml [hunk 5]
```diff
@@ -27,43 +39,85 @@ install:
...
+
+for:
+  -
+    matrix:
+      only:
+        - configuration: Debug
+    build_script:
+      - "echo # > win\\packaging\\CMakeLists.txt"
+      - cmake . -G "%CMAKE_GENERATOR_NAME%"
+          -DCMAKE_BUILD_TYPE=Debug
```
Comment (komainu8) [inline appveyor.yml]: I recommend using RelWithDebInfo in CMAKE_BUILD_TYPE. Because it is optimized and generate debug symbol.
Debug does not optimize so that slow.

Therefore, we should also make configuration of RelWithDebInfo.

Comment (kenhys) [reply appveyor.yml]: I think Debug is enough in this context because this configuration is used for finding build specific issues.
But, in point view of building package, it is better to use same configuration (RelWithDebInfo) because it can reduce tths risk to overloook build error with RelWithDebInfo specific issue.
So I'll fix it.

**Comment by kenhys** [comment]
@komainu8 I've fixed the configuration.
Could you review again, please?

File: appveyor.yml [hunk 6]
```diff
@@ -27,43 +39,85 @@ install:
...
-      -DGRN_WITH_BUNDLED_MECAB=ON
-  - cmake --build . --config Debug
-
 test: off
+
+for:
+  -
+    matrix:
+      only:
+        - configuration: Debug
```
Comment (komainu8) [inline appveyor.yml]: Is modify configuration name to Release or add configuration of Release unnecessary?

Comment (kenhys) [reply appveyor.yml]: Oh, Debug is not appropriate name, so I'll fix it.

**Comment by kenhys** [comment]
@komainu8 
I've fixed configuration name. Is it reasonable?

**Comment by kenhys** [comment]
https://github.com/cosmo0920/PowerShell-for-Mroonga-building/pull/9 had been merged, so
I've fixed not to rewrite versions.ps1. It makes .yml more simple one.
Could you review additional 4 commits in advance, please? (I'm waiting AppVeyor CI result, but It took many hours to finish...)

File: appveyor.yml [hunk 7]
```diff
@@ -116,7 +116,8 @@ for:
       - cd ..\
       - netsh advfirewall firewall add rule name="mariadb" dir=in action=allow protocol=TCP localport=3306 enable=yes profile=any
       - ps: .\install-mroonga.ps1 -mariadbVersion %MARIADB_VERSION% -platforms %PLATFORM%
-      - ps: .\create-mrnzip.ps1 -mariadbVersion %MARIADB_VERSION% -mroongaVersion %MROONGA_VERSION% -platforms %PLATFORM%
+      - ps: $env:ABBREV_COMMIT = $env:APPVEYOR_REPO_COMMIT.Substring(0, 8)
+      - ps: .\create-mrnzip.ps1 -mariadbVersion %MARIADB_VERSION% -mroongaVersion %MROONGA_VERSION%-%ABBREV_COMMIT% -platforms %PLATFORM%
```
[the remark this answers is not in the document]
Comment (kenhys) [reply appveyor.yml]: As a monthly release packages, it is reasonable not to add abbrev commit, as you had pointed out, so I've reverted it.

**Comment by kenhys** [comment]
Now review process by @komainu8  has finished. (https://github.com/mroonga/mroonga/pull/230#discussion_r259213005 commit has reverted)
Wait until AppVeyor results.

**Comment by kenhys** [comment]
After checking https://ci.appveyor.com/project/kenhys/mroonga/builds/22565409, I've found some mistakes, so fixed them.

Remained issues have been solved now! 
https://ci.appveyor.com/project/groonga/mroonga/builds/22595110

**Comment by kenhys** [comment]
@komainu8 Sorry, I've make some mistakes, but these issue was fixed ( https://github.com/mroonga/mroonga/pull/230/commits/2b068b32fa839738d65adac6d46803d677fdad7d ) . Could you confirm it again, please?

**Comment by kenhys** [comment]
@kou  could you review this PR, please?

File: appveyor.yml [hunk 8]
```diff
@@ -4,20 +4,36 @@ notifications:
 
 version: "{build}"
 clone_depth: 10
+configuration:
+  - BuildCheck
+  - ReleasePackage
```
Comment (kou) [inline appveyor.yml]: Can we unify them?
Can we build and archive package in each job like we did in Groonga?

File: appveyor.yml [hunk 9]
```diff
@@ -27,43 +43,81 @@ install:
   - 7z x mariadb-%MARIADB_VERSION%.tar > nul
   - cd mariadb-%MARIADB_VERSION%
   - rmdir /S /Q storage\mroonga\
+  - ps: Get-Content ..\mroonga\version | ForEach-Object { $env:MROONGA_VERSION = $_ }
   - move ..\mroonga storage\mroonga
   - git clone --quiet --depth 1 --recursive https://github.com/groonga/groonga.git ..\groonga
+  - ps: Get-Content ..\groonga\base_version | ForEach-Object { $env:GROONGA_VERSION = $_ }
+  - ps: Get-Content ..\groonga\include\groonga\version.h.in | ForEach-Object { $_ -replace "@GRN_VERSION@","$env:GROONGA_VERSION" } | Set-Content ..\groonga\include\groonga\version.h
```
Comment (kou) [inline appveyor.yml]: This should be generated by CMake.
https://github.com/groonga/groonga/blob/master/CMakeLists.txt#L73-L75

File: appveyor.yml [hunk 10]
```diff
@@ -27,43 +43,81 @@ install:
...
+      - cmake --build . --config RelWithDebInfo
+  -
+    matrix:
+      only:
+        - configuration: ReleasePackage
+    build_script:
+      - set
+      - cd %APPVEYOR_BUILD_FOLDER%\..
+      - git clone --depth 1
+        https://github.com/cosmo0920/PowerShell-for-Mroonga-building.git mrnbuild
```
Comment (kou) [inline appveyor.yml]: This repository is needless when we build our official package by AppVeyor.
Can we move all related staffs to this repository?

File: appveyor.yml [hunk 11]
```diff
@@ -27,43 +43,81 @@ install:
...
+      - cd mrnbuild/powershell
+      - set "PATH=C:\msys64\usr\bin;C:\msys64\mingw64\bin;C:\Windows\System32;C:\Windows;%PATH%"
+      - set MSYSTEM=MINGW64
+      - pacman -Sy --noconfirm bison
+      - ps: .\get-mroonga-batfiles.ps1
```
Comment (kou) [inline appveyor.yml]: We can use files in `packages/windows/*.bat` directly.
Or we can extract command lines in these files here.

File: appveyor.yml [hunk 12]
```diff
@@ -27,43 +43,81 @@ install:
...
+      - ps: .\prepare-building-mroonga.ps1
```
Comment (kou) [inline appveyor.yml]: We don't need this. We already installed Bison manually in the above line.

File: appveyor.yml [hunk 13]
```diff
@@ -27,43 +43,81 @@ install:
...
+      - bison --version
+      - move ..\..\mariadb-%MARIADB_VERSION% work/source
+      - cd .\work
+      - if "%MROONGA_PLATFORM%" == "win32" call .\build-vc2015-zip-32.bat
+      - if "%MROONGA_PLATFORM%" == "winx64" call .\build-vc2015-zip-64.bat
+      - 7z x mariadb-%MARIADB_VERSION%-%MROONGA_PLATFORM%.zip
+      - cd ..\
+      - netsh advfirewall firewall add rule name="mariadb" dir=in action=allow protocol=TCP localport=3306 enable=yes profile=any
+      - ps: .\install-mroonga.ps1 -mariadbVersion $env:MARIADB_VERSION -platforms $env:MROONGA_PLATFORM
+      - ps: .\create-mrnzip.ps1 -mariadbVersion $env:MARIADB_VERSION -mroongaVersion $env:MROONGA_VERSION -platforms $env:MROONGA_PLATFORM
```
Comment (kou) [inline appveyor.yml]: Or we can use 7z like we did in Groonga.

Comment (kou) [reply appveyor.yml]: Should we move archiving command to `on_success`?

File: appveyor.yml (hunk 8 above)
Comment (kenhys) [reply appveyor.yml]: Yes. I've fixed.

File: appveyor.yml (hunk 10 above)
Comment (kenhys) [reply appveyor.yml]: I've imported related stuff.

File: appveyor.yml (hunk 12 above)
Comment (kenhys) [reply appveyor.yml]: I've dropped needless .ps1

File: appveyor.yml (hunk 13 above)
Comment (kenhys) [reply appveyor.yml]: I've moved it.

**Comment by kenhys** [comment]
@kou 
I've fixed based on your feedback!

It is OK for generating artifacts.
https://ci.appveyor.com/project/kenhys/mroonga/builds/22679839

File: appveyor.yml (hunk 11 above)
Comment (kenhys) [reply appveyor.yml]: Fixed

File: appveyor.yml (hunk 9 above)
Comment (kenhys) [reply appveyor.yml]: Fixed

File: appveyor.yml [hunk 14]
```diff
@@ -27,43 +33,42 @@ install:
...
-      -DPLUGIN_XTRADB=NO
-      -DWITH_UNIT_TESTS=OFF
-      -DWITH_MARIABACKUP=OFF
-      -DGRN_WITH_BUNDLED_MECAB=ON
-  - cmake --build . --config Debug
-
 test: off
+
+build_script:
+  - set
```
Comment (kou) [inline appveyor.yml]: Debug print?
Could you remove this?

File: appveyor.yml [hunk 15]
```diff
@@ -27,43 +33,42 @@ install:
...
+  - cd %APPVEYOR_BUILD_FOLDER%\..
+  - set "PATH=C:\msys64\usr\bin;C:\msys64\mingw64\bin;C:\Windows\System32;C:\Windows;%PATH%"
+  - set MSYSTEM=MINGW64
+  - pacman -Sy --noconfirm bison
+  - move mariadb-%MARIADB_VERSION% source
+  - if "%MROONGA_PLATFORM%" == "win32" set MROONGA_BUILD_DIR=build-vc2015-zip-32
+  - if "%MROONGA_PLATFORM%" == "winx64" set MROONGA_BUILD_DIR=build-vc2015-zip-64
+  - mkdir %MROONGA_BUILD_DIR%
+  - cd %MROONGA_BUILD_DIR%
+  - cmake ..\source -G "%CMAKE_GENERATOR_NAME%" -DMRN_GROONGA_EMBED=OFF -DMRN_GROONGA_NORMALIZER_MYSQL_EMBED=OFF -DGRN_WITH_BUNDLED_LZ4=ON -DGRN_WITH_BUNDLED_MECAB=ON
```
Comment (kou) [inline appveyor.yml]: Could you split this long command line to multiple lines like we did in the current  `before_script`?

File: appveyor.yml [hunk 16]
```diff
@@ -27,43 +33,42 @@ install:
...
+  - move .\storage\mroonga\vendor\groonga\include\groonga\version.h ..\source\storage\mroonga\vendor\groonga\include\groonga\
```
Comment (kou) [inline appveyor.yml]: Why is this needed?

File: appveyor.yml [hunk 17]
```diff
@@ -27,43 +33,42 @@ install:
...
+  - cmake --build . --config RelWithDebInfo
+  - cmake --build . --config RelWithDebInfo --target package
+  - move *.zip ..\
+  - cd ..\
+  - 7z x mariadb-%MARIADB_VERSION%-%MROONGA_PLATFORM%.zip
+  - netsh advfirewall firewall add rule name="mariadb" dir=in action=allow protocol=TCP localport=3306 enable=yes profile=any
```
Comment (kou) [inline appveyor.yml]: Could you use multiple lines?

File: appveyor.yml [hunk 18]
```diff
@@ -27,43 +33,42 @@ install:
...
+  - set MROONGA_INSTALL_SCRIPT=.\source\storage\mroonga\packages\windows\install-mroonga.ps1
+  - ps: $command = "$env:MROONGA_INSTALL_SCRIPT -mariadbVersion $env:MARIADB_VERSION -platforms $env:MROONGA_PLATFORM"; Invoke-Expression $command
```
Comment (kou) [inline appveyor.yml]: Could you use multiple lines?

File: appveyor.yml [hunk 19]
```diff
@@ -27,43 +33,42 @@ install:
...
+on_success:
+  - 7z a mariadb-%MARIADB_VERSION%-with-mroonga-%MROONGA_VERSION%-%MROONGA_PLATFORM%.zip mariadb-%MARIADB_VERSION%-%MROONGA_PLATFORM%
```
Comment (kou) [inline appveyor.yml]: Could you use multiple lines?

File: packages/windows/install-mroonga.ps1 [hunk 20]
```diff
@@ -0,0 +1,40 @@
+
```
Comment (kou) [inline packages/windows/install-mroonga.ps1]: Could you remove this empty line?

File: appveyor.yml [hunk 21]
```diff
@@ -27,43 +33,42 @@ install:
...
-      -DPLUGIN_SPIDER=NO
-      -DPLUGIN_TEST_SQL_DISCOVERY=NO
-      -DPLUGIN_TOKUDB=NO
-      -DPLUGIN_XTRADB=NO
-      -DWITH_UNIT_TESTS=OFF
-      -DWITH_MARIABACKUP=OFF
-      -DGRN_WITH_BUNDLED_MECAB=ON
-  - cmake --build . --config Debug
-
 test: off
```
Comment (kou) [inline appveyor.yml]: Could you put `test` section between `build_script` and `on_success`?
The order respects processing order.

File: appveyor.yml [hunk 22]
```diff
@@ -27,43 +33,42 @@ install:
...
+
+build_script:
+  - set
+  - cd %APPVEYOR_BUILD_FOLDER%\..
+  - set "PATH=C:\msys64\usr\bin;C:\msys64\mingw64\bin;C:\Windows\System32;C:\Windows;%PATH%"
+  - set MSYSTEM=MINGW64
+  - pacman -Sy --noconfirm bison
+  - move mariadb-%MARIADB_VERSION% source
+  - if "%MROONGA_PLATFORM%" == "win32" set MROONGA_BUILD_DIR=build-vc2015-zip-32
+  - if "%MROONGA_PLATFORM%" == "winx64" set MROONGA_BUILD_DIR=build-vc2015-zip-64
```
Comment (kou) [inline appveyor.yml]: Can we always use the same build directory?

File: packages/windows/install-mroonga.ps1 [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+    Start-Sleep -m 500
+  } while ($Running)
+}
+
+function Install-Mroonga($mariadbVer, $arch, $installSqlDir) {
+  cd "mariadb-$mariadbVer-$arch"
+  Start-Process .\bin\mysqld.exe
+  Wait-UntilRunning mysqld
+  Get-Content "$installSqlDir\install.sql" | .\bin\mysql.exe -uroot
+  Start-Sleep -m 1000
```
Comment (kou) [inline packages/windows/install-mroonga.ps1]: Is this needed?

File: packages/windows/install-mroonga.ps1 [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+  Wait-UntilTerminate mysqld
+  cd ..
+}
+
+$installSqlDir = ".\share\mroonga"
+
+foreach ($arch in $platforms)
+{
+  Install-Mroonga $mariadbVersion $arch $installSqlDir
+  Start-Sleep -m 500
```
Comment (kou) [inline packages/windows/install-mroonga.ps1]: Is this needed?

File: packages/windows/install-mroonga.ps1 [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+  Get-Content "$installSqlDir\install.sql" | .\bin\mysql.exe -uroot
+  Start-Sleep -m 1000
+  Start-Process .\bin\mysqladmin.exe -ArgumentList "-uroot shutdown"
+  Wait-UntilTerminate mysqld
+  cd ..
+}
+
+$installSqlDir = ".\share\mroonga"
+
+foreach ($arch in $platforms)
```
Comment (kou) [inline packages/windows/install-mroonga.ps1]: Do we need to support multiple platforms?

File: appveyor.yml (hunk 14 above)
Comment (kenhys) [reply appveyor.yml]: Yes, I've fixed.

File: appveyor.yml (hunk 15 above)
Comment (kenhys) [reply appveyor.yml]: Ok, I've fixed.

File: appveyor.yml (hunk 17 above)
Comment (kenhys) [reply appveyor.yml]: Yes, I've fixed.

File: appveyor.yml (hunk 18 above)
Comment (kenhys) [reply appveyor.yml]: I've fixed.

File: appveyor.yml (hunk 21 above)
Comment (kenhys) [reply appveyor.yml]: Ah, I've got it, fixed.

File: appveyor.yml (hunk 22 above)
Comment (kenhys) [reply appveyor.yml]: Surely no need to use other directory here, I've fixed.

File: packages/windows/install-mroonga.ps1 (hunk 24 above)
Comment (kenhys) [reply packages/windows/install-mroonga.ps1]: NO, I've fixed.

File: packages/windows/install-mroonga.ps1 (hunk 23 above)
Comment (kenhys) [reply packages/windows/install-mroonga.ps1]: No, removed it.

File: packages/windows/install-mroonga.ps1 (hunk 25 above)
Comment (kenhys) [reply packages/windows/install-mroonga.ps1]: No. fixed it.

File: appveyor.yml (hunk 19 above)
Comment (kenhys) [reply appveyor.yml]: Yes, i've fixed it.

File: packages/windows/install-mroonga.ps1 (hunk 20 above)
Comment (kenhys) [reply packages/windows/install-mroonga.ps1]: Fixed.

File: appveyor.yml (hunk 16 above)
Comment (kenhys) [reply appveyor.yml]: out of tree build is executed, but there is no need to do so. I've fixed.

**Comment by kenhys** [comment]
@kou 
Thanks.
I've fixed additional your comments.

File: appveyor.yml (hunk 16 above)
Comment (kou) [reply appveyor.yml]: We should support out of tree build.
Maybe, we need to add `"${CMAKE_CURRENT_BUILD_DIR}/${MRN_BUNDLED_GROONGA_RELATIVE_DIR}/include"` to `GROONGA_INCLUDE_DIRS` here: https://github.com/mroonga/mroonga/blob/master/CMakeLists.txt#L292

**Comment by kenhys** [comment]
@kou 
I've fixed to support out-of-tree  build. could you review again, please?

**Comment by kou** [comment]
Thanks!
I've merged.
